Upsell Nudges: Replace site settings Banners with UpsellNudge

sixhours opened the pull request:
#### Changes proposed in this Pull Request

* This PR migrates instances of `Banner` to `UpsellNudge` for consistency, while keeping all Tracks events (impression and click) the same. See #38778
* Also fixes a bug where the JP icon and colors weren't showing up for Jetpack ads.

| Before | After |
| ------- | ------ |
| <img width="1107" alt="Screen Shot 2020-04-09 at 2 41 09 PM" src="https://user-images.githubusercontent.com/2124984/78929874-22aede00-7a71-11ea-96fc-519c49c0cba0.png"> | <img width="1110" alt="Screen Shot 2020-04-09 at 2 41 02 PM" src="https://user-images.githubusercontent.com/2124984/78929888-28a4bf00-7a71-11ea-9272-5c4ada4c1fdf.png"> |
| <img width="768" alt="Screen Shot 2020-04-09 at 2 46 42 PM" src="https://user-images.githubusercontent.com/2124984/78929913-31959080-7a71-11ea-8942-ba3218e0d8f9.png"> | <img width="763" alt="Screen Shot 2020-04-09 at 2 46 30 PM" src="https://user-images.githubusercontent.com/2124984/78929931-38240800-7a71-11ea-9f05-4dca2ebc7ad3.png"> |
| Jetpack <img width="797" alt="Screen Shot 2020-04-09 at 2 56 24 PM" src="https://user-images.githubusercontent.com/2124984/78930580-5cccaf80-7a72-11ea-901d-7cd479d0999b.png"> | Jetpack <img width="766" alt="Screen Shot 2020-04-09 at 2 56 43 PM" src="https://user-images.githubusercontent.com/2124984/78930607-62c29080-7a72-11ea-9e69-68db6a6996ef.png"> |
| WordPress.com
<img width="761" alt="Screen Shot 2020-04-09 at 2 52 31 PM" src="https://user-images.githubusercontent.com/2124984/78930484-373fa600-7a72-11ea-9f93-9984516c1238.png"> | (No visual change) |
| <img width="758" alt="Screen Shot 2020-04-09 at 3 07 15 PM" src="https://user-images.githubusercontent.com/2124984/78931395-e761de80-7a73-11ea-9329-1a5f14e1d607.png"> | <img width="755" alt="Screen Shot 2020-04-09 at 3 07 21 PM" src="https://user-images.githubusercontent.com/2124984/78931416-ed57bf80-7a73-11ea-8df1-daddaef30b9b.png"> |
| WordPress.com 
<img width="1058" alt="Screen Shot 2020-04-10 at 11 10 52 AM" src="https://user-images.githubusercontent.com/2124984/79001241-1e8bca80-7b1c-11ea-8f44-1bc8c352a367.png"> | No visual change |
| Jetpack <img width="1064" alt="Screen Shot 2020-04-10 at 11 11 23 AM" src="https://user-images.githubusercontent.com/2124984/79001264-277c9c00-7b1c-11ea-8593-bb7ec9ed65d5.png"> | Jetpack <img width="1062" alt="Screen Shot 2020-04-10 at 11 11 29 AM" src="https://user-images.githubusercontent.com/2124984/79001290-32cfc780-7b1c-11ea-9d6b-2c647b4c72a0.png"> |
| Jetpack <img width="1068" alt="Screen Shot 2020-04-17 at 1 59 43 PM" src="https://user-images.githubusercontent.com/2124984/79600117-a5a8e780-80b4-11ea-960d-18c5a0f4dba3.png"> | <img width="1081" alt="Screen Shot 2020-04-17 at 2 00 53 PM" src="https://user-images.githubusercontent.com/2124984/79600238-d12bd200-80b4-11ea-9f3f-c4177e7a94b3.png"> |
| WordPress.com <img width="1070" alt="Screen Shot 2020-04-17 at 2 01 02 PM" src="https://user-images.githubusercontent.com/2124984/79600155-b22d4000-80b4-11ea-93a5-381e2ff27493.png"> | No visual change |




#### Testing instructions

Switch to this PR and test the following upsells:

* Go to Marketing -> Traffic on a Jetpack site with the Free plan; note the WordAds upgrade nudge. Double-check Tracks event and its props to make sure there are no changes.
* Go to Settings -> Performance and scroll down to Media on a Jetpack site with the Free plan; note the upgrade nudge, and double-check Tracks events/props.
* Go to Settings -> Performance and look at the Jetpack Search section on a Jetpack site with the free plan. Note the upgrade nudge, double-check Tracks events.
* Go to Settings -> Performance and look at the Jetpack Search section on a WordPress.com site with the free plan. Double-check Tracks events.
* Go to Settings -> Security and scroll down to the Anti-spam heading on a Jetpack site with a free plan. Note the upsell nudge for Akismet and check Tracks events/props.
* Go to Marketing -> Traffic on a Free site and check the upsell under Search engine optimization
* Go to Marketing -> Traffic on a Jetpack site with a free plan and check the upsell under Search engine optimization
* Go to Marketing -> Traffic on a Jetpack site with a free plan and check the upsell under Google Analytics
* Go to Marketing -> Traffic on a WordPress.com site with a free plan and check the upsell under Google Analytics

kwight commented:
Soooo nice to see this getting cleaned up ❤️ 

I do see one upsell go missing on the PR, but it's a WP.com upsell on a JP site, so... not sure why it's there anyways ¯\_(ツ)_/¯

Prod | This branch
--- | ---
![image](https://user-images.githubusercontent.com/349751/79371464-478bd100-7f09-11ea-818a-2e68bc6d3e77.png) | ![image](https://user-images.githubusercontent.com/349751/79371499-54a8c000-7f09-11ea-8296-2b4e7762d9f1.png)

I noticed a query param missing in the second test case (`?feature=video-upload-jetpack-pro` vs. `?feature=video-upload-jetpack-pro&plan=jetpack_premium` on prod), but I think it's related to the https://github.com/Automattic/wp-calypso/pull/41115 that adds in the `href` attribute (so it should fix itself with a rebase). The `plan` query param determines which plan will be highlighted on the page, with a "suggested" badge also displayed.

> the Anti-spam card is completely gone with this PR

I had the anti-spam card, on both this PR and prod (looked good).

sixhours commented:
Putting this back to In Progress until #41143 is shipped

sixhours commented:
Thanks all for your patience in reviewing this one, and helping me untangle upsells! :D

I've fixed:

* The icon/color weirdness with Jetpack ads. Apparently Jetpack used to use plan colors, but I've removed those in favor of the green throughout.
* The "No Ads" ad should only show for WordPress.com sites now, not Jetpack sites.

I haven't been able to reproduce the missing Anti-spam ad. If you start with a fresh test site (I'm using Jurassic Ninja), do you still have this problem?

mmtr commented on client/my-sites/site-settings/form-analytics.jsx:
```diff
@@ -141,6 +141,8 @@ export class GoogleAnalyticsForm extends Component {
 							type: TYPE_PREMIUM,
 							...( siteIsJetpack ? { term: TERM_ANNUALLY } : {} ),
 						} ) }
+						jetpack={ siteIsJetpack }
```
I think this should be `isJetpack`, right?

https://github.com/Automattic/wp-calypso/blob/73d0d63f21ca667747a7f97c534daea652bbdc0a/client/blocks/upsell-nudge/index.jsx#L41

```suggestion
						isJetpack={ siteIsJetpack }
```

mmtr commented on client/my-sites/site-settings/search.jsx:
```diff
@@ -163,6 +163,8 @@ class Search extends Component {
 					event={ 'calypso_jetpack_search_settings_upgrade_nudge' }
 					feature={ FEATURE_SEARCH }
 					plan={ this.props.siteIsJetpack ? PRODUCT_JETPACK_SEARCH : PLAN_BUSINESS }
+					jetpack={ this.props.siteIsJetpack }
```
```suggestion
					isJetpack={ this.props.siteIsJetpack }
```

mmtr commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -376,6 +376,8 @@ export class SeoForm extends React.Component {
 								...( siteIsJetpack ? { term: TERM_ANNUALLY } : {} ),
 							} ) }
+							showIcon={ true }
+							jetpack={ siteIsJetpack }
```
```suggestion
							isJetpack={ siteIsJetpack }
```

sixhours commented:
> Also noted we use the wrong prop for indicating JP sites.

It looks like `Banner` accepts `jetpack`:
https://github.com/Automattic/wp-calypso/blob/7314f49aa90356f5ab5a7f31a7ca238bfdcb3afd/client/components/banner/docs/example.jsx#L67

The input prop name was changed from `jetpack` to `isJetpack` in #41047, but it gets passed to `Banner` as `jetpack` so I think it's correct:
https://github.com/Automattic/wp-calypso/blob/73d0d63f21ca667747a7f97c534daea652bbdc0a/client/blocks/upsell-nudge/index.jsx#L93

I can see how that's confusing, though. I think I'm going to make it `jetpack` again, since that's what has existed in `Banner` until now, and hopefully that will avoid breaking things. 😅

mmtr commented:
@sixhours gotcha, but I think we still need 2 different properties. Currently, `jetpack` is provided by redux and AFAIK we cannot override it from outside. 

Which makes me we wonder: why do we need to pass a property to `UpsellNudge` that indicates if a site is a JP site, if `UpsellNudge` already knows if current site is JP with the redux state?

sixhours commented:
> Re-tested all upsells and got the same results as in #40988 (review).

I spot-checked a bunch of recent PRs on calypso.live and they don't show the upsell on my Jetpack site, either. I'm thinking this is something to do with calypso.live's configuration and not this branch. One of these conditions, maybe?

https://github.com/Automattic/wp-calypso/blob/4537033c6794449ce0a10b69d72daf5fa658d553/client/my-sites/site-settings/spam-filtering-settings.jsx#L55

sixhours commented:
> Which makes me we wonder: why do we need to pass a property to UpsellNudge that indicates if a site is a JP site, if UpsellNudge already knows if current site is JP with the redux state?

...that is a very good point. :D I think I can remove any instances where I pass it to `UpsellNudge` since that was fixed/added directly to the component in #41047.

sixhours commented:
> I had the anti-spam card, on both this PR and prod (looked good).

@kwight Do you remember if you were using calypso.live or your local dev environment when you tested this?

mmtr commented:
> I spot-checked a bunch of recent PRs on calypso.live and they don't show the upsell on my Jetpack site, either. I'm thinking this is something to do with calypso.live's configuration and not this branch.

That did the trick! I got the upsell while testing locally 🙂. Re-verified the rest of them and all good, so this should be good to go 👍
